### PR TITLE
DoRA

### PR DESCRIPTION
**Pull request by calvinpelletier**
#### Context

Adding support for DoRA: https://arxiv.org/abs/2402.09353

Also refactoring `LoRALinear`. The adapter logic is now encapsulated in a submodule `LowRankAdapter` while the base layer logic remains in `LoRALinear`. This lets us just wrap the `LowRankAdapter` module for FSDP, instead of separately wrapping the `lora_a`/`lora_b` projections and the new `magnitude` parameter.

Addresses #1100, #893, #936 

#### Changelog

- [x] Refactor `LoRALinear`/`LowRankAdapter`
- [x] Support DoRA in `LoRALinear` (enabled via `use_dora` in constructor)
- [x] Support DoRA in `get_merged_lora_ckpt`
- [x] Add `use_dora` to all model/component builders
- [x] Example configs for DoRA and QDoRA
- [x] Unit test for DoRA
- [x] Update LoRA recipes so that they notify the model when the base parameters have been loaded (for DoRA initialization)
- [ ] Fix FSDP unit tests
- [ ] Update `recipes/dev/lora_finetune_fsdp2.py`
- [ ] Support DoRA models in state dict conversion PEFT <-> torchtune (`*.lora_magnitude_vector.weight` <-> `*.lora.magnitude`)

#### Test plan

`pytest tests/torchtune/modules/peft/test_lora.py::TestLoRALinear::test_dora`
- checks parity against a reference implementation based on huggingface's
- checks that DoRA initializes correctly
- checks that the merged linear module is approximately the same as the unmerged DoRA linear module 

---

Compared finetunes for LoRA variants (config based on `llama3/8B_lora(_single_device)` for 2 epochs):

LoRA 1xH100
- [final loss: ~1.1](https://api.wandb.ai/links/torchtune/6h3ri4z0)
- time: 325min (~1400 tok/sec)
- GPU mem: 26.8gb

DoRA 1xH100
- [final loss: ~1.1](https://api.wandb.ai/links/torchtune/xp7fpfdu)
- time: 367min (~1300 tok/sec)
- GPU mem: 26.8gb

QLoRA 1xH100
- [final loss: todo](https://api.wandb.ai/links/torchtune/cg856nz6)
- time: todo
- GPU mem: 21.8gb

QDoRA 1xH100
- [final loss: ~1.1](https://api.wandb.ai/links/torchtune/00ggtewj)
- time: 425min (~1100 tok/sec)
- GPU mem: 21.8gb

DoRA 2xH100
- [final loss: ~1.1](https://api.wandb.ai/links/torchtune/gp5t7lyq)
- time: 1656min (~130 tok/sec/gpu)
- GPU mem: 27.1gb + 21.0gb

---

Attempted to replicate section 3.2 ("Weight Decomposition Analysis") of the DoRA paper.

[The code is here.](https://github.com/calvinpelletier/torchtune/blob/aa99fe6ee73468fcecfeea45b6eef2d729b6d0b5/tmp.py)

Results (for comparison with figure 2 of the paper):
Each point is a single query projection (avg direction delta on the X axis, avg magnitude delta on the Y axis).
Full finetune
full
LoRA finetune
lora
DoRA fintune
dora

The difference with the paper is probably due to a different experimental setup (Llama3 finetuned on alpaca instead of VL-BART finetuned on four image-text tasks).

### CODE REVIEW

**Comment by SLR722** [comment]
Thanks for introducing DoRA to torchtune!
- could you add an example distributed yaml config to showcase DoRA? 
- Could you add more details on how you 'checks parity against a reference implementation based on huggingface's' and 'checks that DoRA initializes correctly' in your test plan? 
- In DoRA paper, DoRA outperformances LoRA on several tasks while they didn't mention the loss different. I saw in your comparison, the loss is pretty similar between LoRA and DoRA, could you also eval DoRA and LoRA on some eval benchmarks to see the difference?

File: torchtune/models/convert_weights.py [hunk 1]
```diff
@@ -206,8 +206,8 @@ def _permute(t, n_heads):
 
 # Mapping from torchtune LoRA module names to PEFT LoRA module names
 _TO_PEFT_KEYS = {
-    "lora_a": "lora_A",
-    "lora_b": "lora_B",
+    "lora.a": "lora_A",
```
Comment (SLR722) [inline torchtune/models/convert_weights.py:209]: why do we rename the LoRA keys name?

File: torchtune/modules/peft/lora.py [hunk 2]
```diff
@@ -101,45 +92,112 @@ def _create_weight_and_bias(self):
...
+    def forward(self, x: Tensor, base_out: Tensor, base_weight: Tensor) -> Tensor:
+        if self.disabled:
+            return base_out
+        x = self.dropout(x)
+        lora_out = self.scaling * self.b(self.a(x))
+        if self.use_dora:
+            lora_out = self._dora_forward(x, base_weight, lora_out)
+        return base_out + lora_out
+
+    def _dora_forward(self, x: Tensor, base_weight: Tensor, lora_out: Tensor) -> Tensor:
```
Comment (SLR722) [inline torchtune/modules/peft/lora.py:195]: The DoRA logic looks good to me based on HF's implementation

File: torchtune/modules/peft/lora.py [hunk 3]
```diff
@@ -15,7 +16,7 @@
 from torchtune.modules.peft.peft_utils import AdapterModule
 
 
-class LoRALinear(nn.Module, AdapterModule):
+class LoRALinear(nn.Module):
```
Comment (SLR722) [inline torchtune/modules/peft/lora.py:19]: For this LoRA refactor part, defer to @ebsmothers to check

**Comment by ebsmothers** [comment]
**Update**: I commandeered this PR and pushed some changes with help from @weifengpy. Instead of the previous version where we split out `LoRALinear` into two classes to work with FSDP, I've kept our existing LoRA design and simply added a new `DoRALinear` class that's pretty similar to `LoRALinear` but with the addition of the `magnitude` parameter. 

A lot of the reason for the design of the original version was the usage of FSDP1 APIs, which were a bit too restrictive in terms of sharding params with different values of `requires_grad` in the same `nn.Module`. But with FSDP2 APIs, this is not a problem. However, loading the magnitude param on meta device was a bit tricky.. I wound up using the workaround suggested by Wei in https://github.com/pytorch/pytorch/issues/132721, calling `load_state_dict()` on the `magnitude` params to force the call to `reset_sharded_param` after we initialize them.

There is still some more work to do here though:

(1) Checkpoint save does not work (I think this should be relatively straightforward though)
(2) A lot of cleanup, I'm sure I broke a bunch of tests with my refactor

I think QDoRA will also need more extensive testing too. cc @SLR722

File: torchtune/modules/peft/peft_utils.py [hunk 4]
```diff
@@ -340,3 +365,14 @@ def validate_missing_and_unexpected_for_lora(
                 raise AssertionError(f"Missing LoRA key {k} from adapter state dict")
     if lora_unexpected:
         raise AssertionError("Unexpected key loading adapter")
+
+
+def load_dora_magnitudes(model: nn.Module) -> None:
```
Comment (ebsmothers) [inline torchtune/modules/peft/peft_utils.py:370]: Fyi after https://github.com/pytorch/pytorch/pull/132954/files landed we can now remove this hack by following the example there

File: recipes/dev/lora_finetune_fsdp2.py [hunk 5]
```diff
@@ -314,6 +316,8 @@ def _setup_model(
         for m in reversed(list(model.modules())):
             if isinstance(m, nn.Linear) and m.weight.requires_grad:
                 fully_shard(m, **fsdp_kwargs)
+            if isinstance(m, DoRALinear):
```
Comment (SLR722) [inline recipes/dev/lora_finetune_fsdp2.py:319]: why do we treat DoRALinear separately here? why don't we treat it same as LoRALinear?

File: tests/torchtune/utils/test_distributed.py [hunk 6]
```diff
@@ -194,7 +196,7 @@ def test_lora_fsdp_wrap(self):
 
         adapter_params = get_adapter_params(model)
         set_trainable_params(model, adapter_params)
-        num_lora_ab, num_transformer_layers = _get_n_lora_and_tformer_layers(model)
+        num_lora, num_transformer_layers = _get_n_lora_and_tformer_layers(model)
```
Comment (SLR722) [inline tests/torchtune/utils/test_distributed.py:199]: I saw multiple changes to test_lora_fsdp_wrap test but the unit test doesn't pass on those changes.

Comment (ebsmothers) [reply tests/torchtune/utils/test_distributed.py:199]: Sorry can you elaborate? Do you mean changes when you merge with latest main? Or changes from Calvin's initial version?

File: recipes/dev/lora_finetune_fsdp2.py (hunk 5 above)
Comment (ebsmothers) [reply recipes/dev/lora_finetune_fsdp2.py:319]: I think it's necessary for proper initialization of the magnitude from the sharded LoRA A and B weights. Since otherwise they will be in different FSDP param groups so we would probably need to manually gather and then reshard the weights. You can doublecheck by commenting out L319-320 though to do the usual LoRA sharding, my guess is you will see some kind of DTensor error though.

File: tests/torchtune/utils/test_distributed.py (hunk 6 above)
Comment (SLR722) [reply tests/torchtune/utils/test_distributed.py:199]: There are several changes on test_distributed.py compared with master and I think it's either done by Calvin or your FSDP refactor work. Those tests are broken and I don't have context on why we need these changes

**Comment by SLR722** [comment]
I made several updates to this PR to push it closer to the finish line. 
- fix and clean up several unit tests to let majority of the unit tests pass (there are some remaining issues will list below)
- split test_dora.py and test_lora and improve DoRA logic to make it match with reference implementation 
- fix the checkpointer logic so that dora_distributed, dora_single_device and qdora_single_device can run successfully E2E 

There are some remaining issues may need help from @ebsmothers or other people
- test_dora.py numerical doesn't match ref implementation when dropout p = 0.1 (maybe it's expected because of the non-deterministic of dropout?)
- some tests in test_distributed.py doesn't pass (see the discussions here https://github.com/pytorch/torchtune/pull/1115#discussion_r1725895692)
- 8B_dora_fsdp2.yaml takes long time in get_full_model_state_dict() with >2 ranks which causes NCCL timeout

cc: @ebsmothers

File: tests/torchtune/utils/test_distributed.py (hunk 6 above)
Comment (ebsmothers) [reply tests/torchtune/utils/test_distributed.py:199]: OK yeah I can take a look. The only changes I made were [these ones](https://github.com/pytorch/torchtune/pull/1115/commits/dd694cef19a340a0b181778ae296ee2da5336cc4#diff-a4e7291c0c5acab6a14537dd1c67bf0406396115dafb7459149fe08fd6502758)

Comment (SLR722) [reply tests/torchtune/utils/test_distributed.py]: I fixed the test_distributed.py issue by reverting some of Calvin's change on _distributed.py, which are unnecessary after the DoRA refactor

**Comment by SLR722** [comment]
To further verification the correctness of this DoRA work
- compare the loss between LoRA and DoRA 
  - LoRA: [0.8608](https://wandb.ai/torchtune_dev/torchtune/reports/loss-24-08-26-21-08-15---Vmlldzo5MTU4NTY3)
  - DoRA: [0.8631](https://wandb.ai/torchtune_dev/torchtune/reports/loss-24-08-26-21-09-51---Vmlldzo5MTU4NTg0)

- compare the correlation between changes in direction and magnitude
  - LoRA 
lora
  - DoRA 
dora
   - replicated the statement in Figure 2 in https://arxiv.org/pdf/2402.09353

- run eval on mmlu 
  - LoRA (0.6187)
Screenshot 2024-08-26 at 8 57 48 PM
  - DoRA (0.6214)
Screenshot 2024-08-26 at 9 03 30 PM
  - DoRA has slightly better eval results

File: recipes/configs/dev/llama3/8B_dora_fsdp2.yaml [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+lr_scheduler:
+  _component_: torchtune.modules.get_cosine_schedule_with_warmup
+  num_warmup_steps: 100
+
+loss:
+  _component_: torch.nn.CrossEntropyLoss
+
+# Training
+epochs: 1
+max_steps_per_epoch: 50
```
Comment (ebsmothers) [inline recipes/configs/dev/llama3/8B_dora_fsdp2.yaml:64]: debug code sneaking in 😃

File: torchtune/modules/peft/dora.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+        return weight_norm
+
+    def adapter_params(self) -> List[str]:
+        """
+        Return lora_a.weight and lora_b.weight as adapter params.
+        If bias is enabled, also return lora_a.bias and lora_b.bias.
+        """
+        # NOTE: this function has to be updated if the names of "lora_a" and "lora_b"
+        # in this module change.
+        # TODO: need to add back magnitude, but causing initial check to error out cause it's not defined yet
```
Comment (ebsmothers) [inline torchtune/modules/peft/dora.py:115]: Seems like this can be removed now

File: torchtune/modules/peft/__init__.py [hunk 9]
```diff
@@ -10,21 +10,28 @@
...
     "get_adapter_params",
     "set_trainable_params",
     "validate_missing_and_unexpected_for_lora",
     "validate_state_dict_for_lora",
+    "load_dora_magnitudes",
     "disable_adapter",
     "get_merged_lora_ckpt",
     "get_lora_module_names",
+    "_lora_a_init_params",
+    "_lora_b_init_params",
```
Comment (ebsmothers) [inline torchtune/modules/peft/__init__.py:36]: I don't think we want these to be public

File: tests/torchtune/modules/peft/test_dora.py [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+        mag_norm_scale = (magnitude / weight_norm).view(1, -1)
+        result_dora = (mag_norm_scale - 1) * (
+            F.linear(x, weight)
+        ) + mag_norm_scale * lora_result * self.scaling
+        return result_dora
+
+    def _get_weight_norm(self, weight, lora_weight):
+        weight = weight + self.scaling * lora_weight
+        weight_norm = torch.linalg.norm(weight, dim=1).to(weight.dtype)
+        return weight_norm
```
Comment (ebsmothers) [inline tests/torchtune/modules/peft/test_dora.py:270]: We should refactor this test. In general we try not to put the reference implementation directly in the unit test, but instead use it to determine the expected values, hardcode those in the test, and point to where we got them

File: tests/torchtune/modules/peft/test_dora.py [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+        # Ensure forward passes are the same. This is because DoRALinear should use a special
+        # quantized linear operator that runs compute in higher prec (but only saves the 4 bit quantized tensor)
+        # for autograd.
+        inputs = torch.randn(BSZ, SEQ_LEN, 512, dtype=dtype)
+        torch.manual_seed(0)
+        dora_linear_out = dora_linear(inputs)
+        torch.manual_seed(0)
+        qdora_linear_out = qdora_linear(inputs)
+        torch.testing.assert_close(
+            dora_linear_out, qdora_linear_out, rtol=1e-01, atol=1e-01
```
Comment (SLR722) [inline tests/torchtune/modules/peft/test_dora.py:139]: I need to make this assert very loose since I found to_nf4 will change the original value significantly

File: tests/torchtune/modules/peft/test_dora.py [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+
+from functools import partial
+
+import pytest
+
+import torch
+from tests.test_utils import fixed_init_model
+from torch import nn
+from torchao.dtypes.nf4tensor import NF4Tensor, to_nf4
+from torchtune import utils
```
Comment (SLR722) [inline tests/torchtune/modules/peft/test_dora.py:15]: Note: Need to change here to from torchtune import training after rebasing

File: tests/torchtune/modules/peft/test_dora.py (hunk 10 above)
Comment (SLR722) [reply tests/torchtune/modules/peft/test_dora.py]: Address this comment by only compare the numerical in unit test and move the reference implementation to tests/torchtune/models/llama2/scripts/compare_dora.py

File: tests/torchtune/modules/peft/test_dora.py [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+            DoRALinear(
+                in_dim=512,
+                out_dim=512,
+                rank=RANK,
+                alpha=ALPHA,
+                use_bias=True,
+                quantize_base=False,
+            )
+
+    # @pytest.mark.parametrize("dtype", [torch.bfloat16, torch.float32])
```
Comment (ebsmothers) [inline tests/torchtune/modules/peft/test_dora.py:100]: How come this is commented out?

File: tests/torchtune/modules/peft/test_dora.py (hunk 11 above)
Comment (ebsmothers) [reply tests/torchtune/modules/peft/test_dora.py:139]: Hmm this is a bit suspicious to me. For comparison in the [corresponding LoRA test](https://github.com/pytorch/torchtune/blob/a83eeff0079a73ee04a11e8fc2573ed8f671b231/tests/torchtune/modules/peft/test_lora.py#L144) we do not have to do this. Can we make sure nothing unexpected is happening with the magnitude vector here?

**Review by ebsmothers — COMMENTED** [review]
A few small things, otherwise want to make sure there are no correctness around quantization with the magnitude. After that this looks good to go!

**Review by ebsmothers — APPROVED** [review]
This one is finally ready to go! Huge thanks to both @calvinpelletier for all the initial work and to @SLR722 for taking this one the last mile (really quite a few miles). Really excited to be able to support DoRA!
